src/auth.rs

```diff
@@ -36,20 +36,14 @@ impl TidalCredentials {
             panic!("Application Token needs to be set")
         }
         let token = self.token.to_owned();
-        let mut session_info = match self.session_info {
-            Some(ref session_info) => session_info.to_owned(),
-            None => SessionInfo::new(),
-        };
-
-        let session = session_info.get_session(&token, username, password).await;
+        let session = Session::get_session(&token, username, password).await;
 
-        if let Some(session) = session {
-            session_info = session
+        if session.is_none() {
+            error!("Invalid credentials or Application Token");
+            self
         } else {
-            error!("Invalid credentials or Application Token")
+            self.session(session.unwrap())
```

Comment:
Maybe you could use as return value?
```rust
match session {
  Some(session) => self.session(session),
  None => {
    error!("Invalid credentials or Application Token");
    self
  },
}
```

Reply:
Or maybe even `if let` 🤔  doesn't really matter to be honest. It looks already, just trying to come up with a more idiomatic way to express this 😄